working/macros/feature-specification.md

```diff
@@ -304,20 +304,109 @@ class, allowing you to build up your code fragments however you like.
+with the specified uri and a unique prefix. In the code the identifier will be
+emitted with that unique prefix followed by its simple name.
+
+Note that technically this allows macros to add references to libraries that
+the macro itself does not depend on, and the users application also may not
+depend on. This is discouraged, but not prevented, and should result in an error
+if it happens.
+
+### Generated declarations
+
+A key use of macros is to generate new declarations, and hand-authored code may
+refer to them — it may call macro-generated functions, read macro-generated
+fields, construct macro-generated classes, etc. This means that before macros
+are applied, code may contain identifiers that cannot be resolved. This is not
+an error. Any identifier that can't be resolved before the macro is applied is
+allowed to be resolved to a macro-produced identifier after macros are applied.
+
+Macros are not permitted to introduce declarations that directly conflict with
+existing declarations in the same library. These rules are the same as if the
+code were hand written.
+
+Macros may also end up adding new declarations which shadow existing symbols in
+the library, but don't directly conflict. In this case we want to ensure that
+the intent of any user written code is always clear. Consider the following
+example:
+
+```dart
+int get x => 1;
+
+@generateX 
+class Bar {
+  // Generated: int get x => 2;
+
+  // Should this return the top level `x`, or the generated instance getter?
+  int get y => x; 
+}
+```
+
+There are several potential choices to we could make here:
+
+1.  We could say that any identifier that can be resolved before macro
+    application keeps its original resolution (so `x` would still resolve to the
+    original, top level `x`).
+2.  We could re-resolve all identifiers after the macros are applied, which can
+    possibly change what they resolve to (in this case `x` would resolve to the
+    generated instance getter `x`).
+3.  We could make it some kind of error for a macro to introduce an identifier
+    that shadows another.
+4.  We could make it error to use an identifier shadowed by one produced by a
+    macro.
+
+The first two choices could be very confusing to users, some will expect one
+behavior while others expect the other. The third choice would work but might be
+overly restrictive. The final option still avoids the ambiguity, and is a bit
+more permissive than the third.
+
+It similarly also is not allowed for one macro to produce an declaration that
+the identifier resolves to and then another macro to produce another declaration
+that then shadows that one. In other words, any hand-authored identifier may be
+resolved at any point during macro application, but it may only be resolved
+once.
+
+At the same time, a macro application in a library must be able to shadow
+identifiers in other libraries, in order to allow for modular compilation. It
```

Comment:
I'm not sure what this means, I don't know if "modular compilation" (an implementation technique) is the right thing to focus on. Are you trying to say that a macro application can produce a declaration that shadows an imported one?

Reply:
What I am trying to say is that if I import a library, macros in that imported library *are* allowed to change the resolution of identifiers in my library. Otherwise my library can cause a compile time error in a library I import, which is bad. And in general that sounds like a very expensive thing to check for.

Case in point:

```dart
import 'a.dart'; // has a macro generated top level getter `x`

class A {
  int get x => 2;
}

class B {
  // Which `x`? Its actually the one that comes from `a.dart`, because top level getters
  // take precedence over super getters.
  int get y => x; 
}

void main() {
  print(B().y); // 1
}
```

I don't know if there are more situations than just this that exist, this is just kind of a weird quirk, but it does exist.

Reply:
Ah, I get it. Yes, I was just taking for granted that all macro applications in imported libraries are done before the import is processed at all. Once you import a library, you can't tell what parts of it came from macros and what didn't.

I think there's probably a better way to express this. Maybe early in the entire section talk about this is how scoping works *within* a single library and explicitly state that macro applications are handled separately for each library. Or maybe that's implicit given the ordering section? Maybe you just need to say "Here is how scoping works within a library."?

Reply:
I added a part above that clarifies the rules only apply within a single library.